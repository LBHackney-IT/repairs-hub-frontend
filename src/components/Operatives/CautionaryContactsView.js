import BackButton from '../Layout/BackButton'
import { Table, THead, TBody, TR, TH, TD } from '../Layout/Table'
import { CAUTIONARY_CONTACT } from '../../utils/cautContactAlertsDescription'
const CautionaryContactView = () => (
  <>
    <BackButton />
    <h3 className="lbh-heading-h3">Cautionary contact</h3>
    <Table>
      <THead>
        <TR className="lbh-body">
          <TH scope="col">Code</TH>
          <TH scope="col">Description</TH>
        </TR>
      </THead>
      <TBody>
        {CAUTIONARY_CONTACT.map((entry, index) => (
          <TR key={index} index={index}>
            <TD>{entry.code}</TD>
            <TD>{entry.description}</TD>
          </TR>
        ))}
      </TBody>
    </Table>
  </>
)

export default CautionaryContactView
