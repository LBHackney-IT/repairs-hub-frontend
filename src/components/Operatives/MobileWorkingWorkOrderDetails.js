import PropTypes from 'prop-types'
import { WorkOrder } from '@/models/workOrder'
import { GridColumn, GridRow } from '../Layout/Grid'
import { useRouter } from 'next/router'
import { getCautionaryAlertsType } from '@/utils/cautionaryAlerts'
import TruncateText from '../Layout/TruncateText'

const MobileWorkingWorkOrderDetails = ({
  property,
  workOrder,
  personAlerts,
  locationAlerts,
}) => {
  const router = useRouter()
  const cautionaryAlertsType = getCautionaryAlertsType(
    locationAlerts,
    personAlerts
  )

  const cautContactURL = () => {
    router.push({
      pathname: `/work-orders/cautionary-alerts`,
      query: {
        cautContactCodes: cautionaryAlertsType,
      },
    })
  }

  return (
    <>
      <div className="operative-work-order">
        <h1 className="lbh-heading-h1">
          WO&nbsp;{workOrder.reference.toString().padStart(8, '0')}
        </h1>
        <div className="priority-text govuk-!-margin-top-0">
          {workOrder.isHigherPriority() ? (
            <div className="text-dark-red lbh-heading-h3">
              {' '}
              {workOrder.priority.toLowerCase().split(' ').slice(-1)}
            </div>
          ) : (
            <div className="lbh-heading-h3">
              {' '}
              {workOrder.priority.toLowerCase().split(' ').slice(-1)}
            </div>
          )}
        </div>
        <div className="lbh-heading-h4">Description</div>
        <TruncateText
          text={workOrder.description}
          numberOfLines="3"
          pTagClassName={'govuk-body govuk-!-margin-bottom-0'}
          linkClassName={'govuk-body'}
        ></TruncateText>

        {workOrder.plannerComments && (
          <>
            <div className="lbh-heading-h5">Planner Comment</div>
            <p className="govuk-body">{workOrder.plannerComments}</p>
          </>
        )}
        <div className="work-order-information">
          {cautionaryAlertsType && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <a
                  className="lbh-heading-h5 lbh-link"
                  href="#"
                  id="caut-alerts"
                  onClick={cautContactURL}
                >
                  Caut. alerts
                </a>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="govuk-warning-text lbh-warning-text">
                  <span
                    className="govuk-warning-text__icon person-alert--icon"
                    aria-hidden="true"
                  >
                    !
                  </span>
                  <strong className="govuk-warning-text__text person-alert--text">
                    {cautionaryAlertsType}
                  </strong>
                </div>
              </GridColumn>
            </GridRow>
          )}

          <GridRow className="align-grid-row">
            <GridColumn width="one-half">
              <div className="lbh-body property-name">Address</div>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="lbh-body">
                {property.address.shortAddress} {property.address.postalCode}
              </div>
            </GridColumn>
          </GridRow>

          {workOrder.callerName && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Contact</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.callerName}</div>
              </GridColumn>
            </GridRow>
          )}

          {workOrder.callerNumber && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Telephone no</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.callerNumber}</div>
              </GridColumn>
            </GridRow>
          )}

          {workOrder.appointment && workOrder.appointment.note && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Comment</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.appointment.note}</div>
              </GridColumn>
            </GridRow>
          )}
        </div>
      </div>
    </>
  )
}

MobileWorkingWorkOrderDetails.propTypes = {
  property: PropTypes.object.isRequired,
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
  personAlerts: PropTypes.array.isRequired,
  locationAlerts: PropTypes.array.isRequired,
}

export default MobileWorkingWorkOrderDetails
