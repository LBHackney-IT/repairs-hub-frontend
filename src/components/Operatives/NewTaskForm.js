import { useForm } from 'react-hook-form'
import { useState, useEffect } from 'react'
import { useRouter } from 'next/router'
import { frontEndApiRequest } from '@/utils/frontEndApiClient/requests'
import { PrimarySubmitButton } from '../Form'
import BackButton from '../Layout/BackButton'
import RateScheduleItem from '../WorkElement/RateScheduleItem'
import Spinner from '../Spinner'
import ErrorMessage from '../Errors/ErrorMessage'
import { buildVariationFormData } from '@/utils/hact/jobStatusUpdate/variation'
import { getWorkOrder } from '../../utils/requests/workOrders'
import { APIResponseError } from '../../types/requests/types'
import { formatRequestErrorMessage } from '../../utils/errorHandling/formatErrorMessage'

const NewTaskForm = ({ workOrderReference }) => {
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState()
  const [currentUser, setCurrentUser] = useState({})
  const [sorCodes, setSorCodes] = useState([])
  const [existingTasks, setExistingTasks] = useState({})
  const { register, handleSubmit, errors } = useForm()
  const router = useRouter()

  const getNewTaskForm = async (reference) => {
    setError(null)

    try {
      const workOrderResponse = await getWorkOrder(reference)

      if (!workOrderResponse.success) {
        throw workOrderResponse.error
      }

      const workOrder = workOrderResponse.response

      const sorCodes = await frontEndApiRequest({
        path: '/api/schedule-of-rates/codes',
        method: 'get',
        params: {
          tradeCode: workOrder.tradeCode,
          propertyReference: workOrder.propertyReference,
          contractorReference: workOrder.contractorReference,
          showAdditionalTrades: true,
        },
      })

      setSorCodes(sorCodes)

      const currentUser = await frontEndApiRequest({
        method: 'get',
        path: '/api/hub-user',
      })

      setCurrentUser(currentUser)

      const tasksAndSors = await frontEndApiRequest({
        method: 'get',
        path: `/api/workOrders/${workOrderReference}/tasks`,
      })

      setExistingTasks(tasksAndSors)
    } catch (e) {
      setSorCodes([])

      if (e instanceof APIResponseError) {
        setError(e.message)
      } else {
        setError(formatRequestErrorMessage(e))
      }
    }

    setLoading(false)
  }

  const onFormSubmit = async (e) => {
    setLoading(true)

    const newTask = [
      {
        ...e.rateScheduleItems.operative,
        description: e.rateScheduleItems.operative.code.split(' - ')[1],
      },
    ]

    const workOrderUpdateFormData = buildVariationFormData(
      existingTasks,
      newTask,
      workOrderReference
    )

    try {
      await frontEndApiRequest({
        method: 'post',
        path: `/api/jobStatusUpdate`,
        requestData: workOrderUpdateFormData,
      })

      router.push(
        `/operatives/${currentUser?.operativePayrollNumber}/work-orders/${workOrderReference}`
      )
    } catch (e) {
      console.error(e)

      setError(formatRequestErrorMessage(e))
    }

    setLoading(false)
  }

  useEffect(() => {
    setLoading(true)
    setError(null)

    getNewTaskForm(workOrderReference)
  }, [])

  return (
    <>
      {loading ? (
        <Spinner />
      ) : (
        <>
          <BackButton />
          {sorCodes && (
            <>
              <h1 className="lbh-heading-h2 govuk-!-margin-bottom-4">
                New SOR
              </h1>
              <form
                role="form"
                id="repair-request-form"
                onSubmit={handleSubmit(onFormSubmit)}
              >
                <RateScheduleItem
                  sorCodes={sorCodes}
                  register={register}
                  showRemoveRateScheduleItem={false}
                  removeRateScheduleIte={false}
                  errors={errors}
                  index={'operative'}
                />
                <div className="button-pair">
                  <PrimarySubmitButton
                    id="submit-work-order-task-add"
                    label="Confirm"
                  />
                </div>
              </form>
            </>
          )}
          {error && <ErrorMessage label={error} />}
        </>
      )}
    </>
  )
}

export default NewTaskForm
