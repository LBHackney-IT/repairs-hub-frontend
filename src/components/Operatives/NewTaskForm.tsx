import WorkOrderUpdateView from '../WorkOrder/Update'
import { mobileViewUpdateWorkOrderLinks } from '../../utils/successPageLinks'

interface Props {
  workOrderReference: string
  operativePayrollNumber: string
}

const NewTaskForm = (props: Props) => {
  const { workOrderReference, operativePayrollNumber } = props

  return (
    <WorkOrderUpdateView
      reference={workOrderReference}
      // isMobileView={true}
      mobileViewLinks={mobileViewUpdateWorkOrderLinks(
        operativePayrollNumber,
        workOrderReference
      )}
    />
  )
}

export default NewTaskForm
