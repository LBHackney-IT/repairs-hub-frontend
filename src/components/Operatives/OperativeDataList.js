import PropTypes from 'prop-types'
import { DataList } from '../Form'

const OperativeDataList = ({
  selectedOperative,
  name,
  options,
  label,
  index,
  register,
  errors,
  isOperativeNameSelected,
  getValues,
}) => {
  const onChange = () => {
    isOperativeNameSelected(true)
  }

  return (
    <>
      <DataList
        name={name}
        onChange={onChange}
        defaultValue={selectedOperative}
        options={options}
        label={label}
        register={register({
          validate: (value) => {
            if (!options.includes(value)) {
              return 'Please select an operative'
            } else if (
              Object.entries(getValues())
                .filter(
                  ([k]) => k !== `operative-${index}` && k.match(/operative-/)
                )
                .some(([, name], i) => name === value && i < index)
            ) {
              return 'This operative has already been added'
            }
          },
        })}
        error={errors && errors[name]}
        additionalDivClasses={['govuk-!-display-inline-block']}
      />
    </>
  )
}

OperativeDataList.propTypes = {
  name: PropTypes.string.isRequired,
  selectedOperative: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  options: PropTypes.array.isRequired,
  register: PropTypes.func.isRequired,
}

export default OperativeDataList
