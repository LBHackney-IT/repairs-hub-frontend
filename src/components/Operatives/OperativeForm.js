import PropTypes from 'prop-types'
import { useForm } from 'react-hook-form'
import { PrimarySubmitButton } from '../Form'
import BackButton from '../Layout/BackButton'
import SelectOperatives from './SelectOperatives'

const OperativeForm = ({
  onSubmit,
  assignedOperativesToWorkOrder,
  availableOperatives,
  selectedPercentagesToShowOnEdit,
  totalSMV,
}) => {
  const { handleSubmit, register, errors, trigger, getValues } = useForm({})

  return (
    <>
      <div>
        <BackButton />

        <h2 className="lbh-heading-h2">Add operative</h2>

        <form role="form" onSubmit={handleSubmit(onSubmit)}>
          <SelectOperatives
            assignedOperativesToWorkOrder={assignedOperativesToWorkOrder}
            availableOperatives={availableOperatives}
            register={register}
            errors={errors}
            selectedPercentagesToShowOnEdit={selectedPercentagesToShowOnEdit}
            trigger={trigger}
            getValues={getValues}
            totalSMV={totalSMV}
          />

          <PrimarySubmitButton label={'Confirm'} />
        </form>
      </div>
    </>
  )
}

OperativeForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  totalSMV: PropTypes.number.isRequired,
}

export default OperativeForm
