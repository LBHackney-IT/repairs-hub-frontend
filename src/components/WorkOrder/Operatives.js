const Operatives = ({ operatives }) => {
  return (
    <p className="lbh-body-xs">
      {`${operatives.length > 1 ? 'Operatives' : 'Operative'}: ${operatives
        .map((operative) => operative.name)
        .join(', ')}`}
    </p>
  )
}

export default Operatives
