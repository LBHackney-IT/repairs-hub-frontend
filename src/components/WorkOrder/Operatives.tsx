import { Operative } from '../../models/operativeModel'

interface Props {
  operatives: Operative[]
}

const Operatives = ({ operatives }: Props) => {
  return (
    <p className="lbh-body-xs">
      {`${operatives.length > 1 ? 'Operatives' : 'Operative'}: ${operatives
        .map((operative) => operative.name)
        .join(', ')}`}
    </p>
  )
}

export default Operatives
