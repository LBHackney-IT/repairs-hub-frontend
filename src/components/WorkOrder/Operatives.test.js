import { render } from '@testing-library/react'
import Operatives from '@/components/WorkOrder/Operatives'

describe('Operatives component', () => {
  it('Renders a string with a prefix of "Operative" when passed one operative', () => {
    const { asFragment } = render(
      <Operatives operatives={[{ name: 'Operative Name 1' }]} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('Renders a string with a prefix of "Operatives" when passed multiple operative', () => {
    const { asFragment } = render(
      <Operatives
        operatives={[
          { name: 'Operative Name 1' },
          { name: 'Operative Name 2' },
        ]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
