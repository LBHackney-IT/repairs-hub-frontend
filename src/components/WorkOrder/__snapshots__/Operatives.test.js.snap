// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operatives component Renders a string with a prefix of "Operative" when passed one operative 1`] = `
<DocumentFragment>
  <p
    class="lbh-body-xs"
  >
    Operative: Operative Name 1
  </p>
</DocumentFragment>
`;

exports[`Operatives component Renders a string with a prefix of "Operatives" when passed multiple operative 1`] = `
<DocumentFragment>
  <p
    class="lbh-body-xs"
  >
    Operatives: Operative Name 1, Operative Name 2
  </p>
</DocumentFragment>
`;
